main/src/main/scala/sbt/internal/EvaluateConfigurations.scala

```diff
@@ -151,10 +151,13 @@ private[sbt] object EvaluateConfigurations {
     val allGeneratedFiles = (definitions.generated ++ dslEntries.flatMap(_.generated))
     loader =>
       {
-        val projects =
-          definitions.values(loader).collect {
-            case p: Project => resolveBase(file.getParentFile, p)
+        val projects = {
+          val compositeProjects = definitions.values(loader).flatMap {
+            case p: CompositeProject => p.componentProjects.map(resolveBase(file.getParentFile, _))
+            case _                   => Nil
           }
+          CompositeProject.uniqueId(compositeProjects)
```

Comment:
If you first `flatMap`, *then* call `uniqueId`, you have already lost the distinction between things that were in the original list versus those added via the `componentProjects`. That can't properly work: they will *all* be kept.

Reply:
It should be possible to *first* call `CompositeProject.expand(definitions.values(loader))`, then `map` the result with `.map(resolveBase(file.getParentFile, _))`.